MAX32xxx One-Wire Driver

ozersa opened the pull request:
One-Wire (1W) driver for MAX32xxx MCUs,

| Has 1-W       |  Not have 1-W  |
| -------------- | ---------------- |
| MAX32655    | MAX32662      |
| MAX32666    | MAX32670      |
| MAX32680    | MAX32672      |
| MAX32690    | MAX32675      |


MAX32655FTHR board does not have 1W pinout so that is is not enabled for this board.

Tested with:
[tests/drivers/w1/w1_api](https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/drivers/w1/w1_api)
![image](https://github.com/zephyrproject-rtos/zephyr/assets/46590392/fa96c559-0f48-4fa3-9b1e-93a43b3f9c8f)

[samples/drivers/w1/scanner](https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/drivers/w1/scanner)
Test output for no 1W slave on the bus
![image](https://github.com/zephyrproject-rtos/zephyr/assets/46590392/dee879c2-98f7-4683-a23e-0f806b09115f)

Test output for 1W slave on the bus (Added [MAX31825](https://www.analog.com/en/products/max31825.html) 1W slave on the bus)
![image](https://github.com/zephyrproject-rtos/zephyr/assets/46590392/2d2a0dac-5160-4274-91af-b2d288549037)

ozersa committed:
drivers: w1: Add MAX32xxx 1-Wire driver

Added 1-Wire master driver for MAX32xxx MCUs

Signed-off-by: Sadik Ozer <sadik.ozer@analog.com>

ozersa committed:
dts: Add 1-Wire inside device tree

Add 1-Wire register file
Some MAX32 MCUs has 1-Wire peripheral some one not
So that added in device related dtsi file.

Has 1-W       Not have 1-W
MAX32655        MAX32662
MAX32666        MAX32670
MAX32680        MAX32672
MAX32690        MAX32675

Signed-off-by: Sadik Ozer <sadik.ozer@analog.com>

ozersa committed:
boards: adi: Add 1-Wire driver in MAX32655EVKIT driver list

1-Wire interface can be used with MAX32655EVKIT
It added in supported interface list

Signed-off-by: Sadik Ozer <sadik.ozer@analog.com>

ozersa committed:
tests: drivers: w1: Add MAX32655EVKIT board overlay file

Add MAX32655EVKIT board overlay files to w1_api test be supported.

Signed-off-by: Sadik Ozer <sadik.ozer@analog.com>

ozersa committed:
boards: adi: Add 1-Wire driver in MAX32666 driver list

1-Wire interface can be used with MAX32666 boards
It added in supported interface list

Signed-off-by: Sadik Ozer <sadik.ozer@analog.com>

ozersa committed:
tests: drivers: w1: Add MAX32666FTHR board overlay file

Add MAX32666FTHR board overlay files to w1_api test be supported.

Signed-off-by: Sadik Ozer <sadik.ozer@analog.com>

ozersa committed:
tests: drivers: w1: Add MAX32690EVKIT board overlay file

Add MAX32690EVKIT board overlay files to w1_api test be supported.

Signed-off-by: Sadik Ozer <sadik.ozer@analog.com>

hfakkiz committed:
tests: drivers: w1: Enable 1-wire driver tests for max32666evkit board

Enable 1-wire driver test for MAX32666EVKIT

Signed-off-by: Furkan Akkiz <hasanfurkan.akkiz@analog.com>

ttmut committed:
tests: drivers: w1: Add MAX32680EVKIT board overlay file

Support MAX32680EVKIT in 1-wire tests.

Signed-off-by: Tahsin Mutlugun <Tahsin.Mutlugun@analog.com>

ozersa committed:
boards: adi: Add 1-Wire driver in MAX32680EVKIT driver list

Add 1W driver in MAX32680EVKIT driver list

Signed-off-by: Sadik Ozer <sadik.ozer@analog.com>

ozersa committed:
boards: adi: Add 1-Wire driver in MAX32690EVKIT driver list

Add 1W driver in MAX32690EVKIT driver list

Signed-off-by: Sadik Ozer <sadik.ozer@analog.com>

ozersa committed:
boards: adi: Add 1-Wire driver in APARD32690 driver list

Add 1W driver in APARD32690 driver list

Signed-off-by: Sadik Ozer <sadik.ozer@analog.com>

ozersa committed:
tests: drivers: w1: Add APARD32690 board overlay file

Add APARD32690 board overlay files to w1_api test be supported.

Signed-off-by: Sadik Ozer <sadik.ozer@analog.com>

gmarull commented on dts/bindings/w1/adi,max32-w1.yaml:
```diff
@@ -0,0 +1,61 @@
+  interrupts:
+    required: true
+
+  pinctrl:
+    type: phandles
+
```
wrong?

ozersa replied:
Updated, thanks.

decsny commented on dts/bindings/w1/adi,max32-w1.yaml:
```diff
@@ -0,0 +1,64 @@
+  long-line-mode:
+    required: true
+    type: int
+    enum: [0, 1]
```
suggestion is to use string enum but not required

ozersa replied:
Thanks for the feedback, I tried to match it with UserGuide, it might be more clear if we keep it same as UG.
![image](https://github.com/zephyrproject-rtos/zephyr/assets/46590392/6c3b7b62-10e7-49a9-9f0b-14f7350bd7a4)

ozersa replied:
Update: Converted to boolean as per of @str4t0m feedback. Please see: https://github.com/zephyrproject-rtos/zephyr/pull/75009#discussion_r1730874035

MaureenHelm reviewed (commented):
Can you also update `samples/drivers/w1/scanner/sample.yaml` and `tests/drivers/w1/w1_api/testcase.yaml` to ensure that twister exercises them on the max32 boards? They both currently use `platform_allow` to constrain to a small list of boards. Need to extend the list or remove it entirely.

ozersa commented:
> Can you also update `samples/drivers/w1/scanner/sample.yaml` and `tests/drivers/w1/w1_api/testcase.yaml` to ensure that twister exercises them on the max32 boards? They both currently use `platform_allow` to constrain to a small list of boards. Need to extend the list or remove it entirely.

Added for tests/drivers/w1/w1_api. 

Regarding sample, I have removed this addition, due to [MAX31825](https://www.analog.com/en/products/max31825.html) One-Wire slave sensor not onboard component, it is external evkit. I will revisit this sample after push [MAX31825](https://www.analog.com/en/products/max31825.html) sensor driver to the zephyr.

ozersa commented:
Conflicts fixed and rebased onto main.

str4t0m commented:
> Conflicts fixed and rebased onto main.
Thanks, I'll have a closer look at the PR this week.

ozersa commented:
Rebased onto main to fix conflicts.

ozersa commented:
Will be rebased after below PR being merged.
https://github.com/zephyrproject-rtos/zephyr/pull/73560
https://github.com/zephyrproject-rtos/zephyr/pull/76593

MaureenHelm commented:
Needs another rebase

ozersa commented:
Rebased onto main, and conflicts are fixed.

MaureenHelm commented on boards/adi/max32690evkit/max32690evkit_max32690_m4.yaml:
```diff
@@ -15,5 +15,6 @@ supported:
   - trng
   - dma
   - watchdog
+  - 1w
```
w1

ozersa replied:
Thanks, fixed.

ttmut commented on drivers/w1/w1_max32.c:
```diff
@@ -0,0 +1,207 @@
+	return ret;
+}
+
+static int api_write_byte(const struct device *dev, uint8_t byte)
```
You could change the prefix of `api_` functions to `w1_max32_`. I had a similar feedback in the past.

https://github.com/zephyrproject-rtos/zephyr/pull/76627#discussion_r1702223265

ozersa replied:
Thanks but, there are too many w1 and max32 inside the code, so we decided to add api_ prefix for all drivers to get more readable driver code file.
Existing uart/ gpio... has same format, I prefer to keep it as it is.

ttmut commented on tests/drivers/w1/w1_api/testcase.yaml:
```diff
@@ -7,6 +7,10 @@ common:
+    - max32655evkit/max32655/m4
+    - max32680evkit/max32680/m4
+    - max32690evkit/max32690/m4
+    - apard32690/max32690/m4
```
Why are MAX32666 boards not in this list?

ozersa replied:
Because I forget to add it.

ttmut replied:
max32666evkit is still missing it seems.

ozersa replied:
Thanks added.

ttmut reviewed (commented):
I have given some minor feedback you might want to consider.

ozersa commented on drivers/w1/w1_max32.c:
```diff
@@ -0,0 +1,207 @@
+	const struct max32_w1_config *cfg = dev->config;
+	mxc_owm_regs_t *regs = cfg->regs;
+
+	/* 0 if no 1-wire devices reponded during the presence pulse, 1 otherwise */
```
Thanks

ozersa commented on drivers/w1/w1_max32.c:
```diff
@@ -0,0 +1,207 @@
+	ret = MXC_OWM_Reset();
+	if (ret > 0) {
+		/* Check OW pin input state due to presence detect pin seems not work well */
+		if (regs->ctrl_stat & MXC_F_OWM_CTRL_STAT_OW_INPUT) {
+			ret = 1; /* At least 1 device on the line */
+		} else {
+			ret = 0; /* no device on the line */
+		}
```
Thanks, I prefer existing one due to it is more readable.

ozersa commented on drivers/w1/w1_max32.c:
```diff
@@ -0,0 +1,207 @@
+static int api_reset_bus(const struct device *dev)
+{
+	int ret;
+	const struct max32_w1_config *cfg = dev->config;
```
Added.

ozersa commented on tests/drivers/w1/w1_api/boards/max32655evkit_max32655_m4.overlay:
```diff
@@ -0,0 +1,24 @@
+	internal-pullup = <1>;
+	external-pullup = <0>;
+	long-line-mode = <0>;
+};
```
It is because of number of limited 1W GPIOs (2 pins)  and multiple available alternate functions (see below table for a device) on EVKITs, 
So it might not be good option to enforce a default 1W pins for EVKITs... Anyway both option has pros and cons...
![image](https://github.com/user-attachments/assets/0b5e059e-d13b-420a-abcf-3aa3536c62fd)

Moved and removed power selection which are not necessary for the w1 test.

str4t0m commented on dts/bindings/w1/adi,max32-w1.yaml:
```diff
@@ -0,0 +1,64 @@
+      1 - Pullup pin is active low when enabled.
+      2 - Pullup pin is not used for an external pullup.
+
+  long-line-mode:
+    required: true
+    type: int
+    enum: [0, 1]
```
I too think that it would align better with expectations of dts configurations in zephyr, if that would be a boolean instead.

ozersa replied:
Boolean make sense, added, thanks.

str4t0m commented on tests/drivers/w1/w1_api/testcase.yaml:
```diff
@@ -7,6 +7,12 @@ common:
   platform_allow:
     - nucleo_g0b1re
     - nrf52840dk/nrf52840
+    - max32655evkit/max32655/m4
+    - max32680evkit/max32680/m4
+    - max32690evkit/max32690/m4
+    - apard32690/max32690/m4
+    - max32666fthr/max32666/cpu0
+    - max32666evkit/max32666/cpu0
   integration_platforms:
     - nrf52840dk/nrf52840
 
```
For the integrated w1 peripheral we should add an additional test configuration. Something like drivers.w1.max32 and for that test you could add the filter filter: dt_compat_enabled("adi,max32-w1")

Regarding the test configuration. Could you please make an additional change:
Could you move the existing platform_allow part just to the drivers.w1.w1-serial. I have added it, as I can't test all platforms and not all platforms might allow the required configuration of the peripheral.
And at the moment I am not sure about the overriding behavior of integration_platforms. Either leave it here or add the nrf52840dk to each test config but the max32 test and restrict/not-restrict the max32 test to the preference of ADI platforms.

ozersa replied:
Please take a look: https://github.com/zephyrproject-rtos/zephyr/pull/75009/commits/d6b2c821ee92a38db46873a222e2806b0e4656cf

str4t0m commented on drivers/w1/w1_max32.c:
```diff
@@ -0,0 +1,207 @@
+{
+	int ret;
+
+	ret = MXC_OWM_ReadBit();
+	if (ret < 0) {
+		if (MXC_OWM_GetPresenceDetect() == 0) {
+			/* if no slave connected to the bus, read bits shall be logical ones */
+			ret = 1;
+		} else {
+			return -EIO;
+		}
```
I'm trying to understand the behaviour we will have here without having access to the hardware:
Looking at the read_bit implementation of owm_reva the only time a negative value is returned in case of a short on the line. There are a few details I can't derive from the datasheet. Assuming that short was already present in the last reset, would the present flag then also always be set?

ozersa replied:
Yes, if the line being short it return negative value, https://github.com/zephyrproject-rtos/hal_adi/blob/main/MAX/Libraries/PeriphDrivers/Source/OWM/owm_reva.c#L156
The present flag being clear in that case, https://github.com/zephyrproject-rtos/hal_adi/blob/main/MAX/Libraries/PeriphDrivers/Source/OWM/owm_me18.c#L92C25-L92C35

str4t0m commented on drivers/w1/w1_max32.c:
```diff
@@ -0,0 +1,207 @@
+	mxc_owm_regs_t *regs = cfg->regs;
+
+	/* 0 if no 1-wire devices responded during the presence pulse, 1 otherwise */
+	ret = MXC_OWM_Reset();
+	if (ret > 0) {
+		/* Check OW pin input state due to presence detect pin seems not work well */
+		if (regs->ctrl_stat & MXC_F_OWM_CTRL_STAT_OW_INPUT) {
+			ret = 1; /* At least 1 device on the line */
+		} else {
+			ret = 0; /* no device on the line */
+		}
```
Having a short look at the datasheet, and the hal. It seems `MXC_OWM_Reset()` waits until the reset is complete and returns the result of the stat presence detect flag. Which sounds reasonable. Is there a known issue/errata on a few socs or do your tests show that relying on the presence detect flag directly is not sufficient?

str4t0m replied:
I see thanks. Was it something with the zephyr w1_api tests or something you observed with the temperature driver?
I am asking as the w1 api test does not test very much, it tests a bit more than that everything compiles.

ozersa replied:
Added these logic to it pass w1_api tests. Otherwise it was failed while executing some tests that execute reset.
![image](https://github.com/user-attachments/assets/02005f35-02c3-4c89-ada2-2b470dbb25d0)

str4t0m replied:
I have bought a board and finally got it delivered so I will have a closer look at this some time this week.

str4t0m commented on drivers/w1/w1_max32.c:
```diff
@@ -0,0 +1,207 @@
+	const struct pinctrl_dev_config *pctrl;
+	const struct device *clock;
+	struct max32_perclk perclk;
+	int internal_pullup;
+	int external_pullup;
+	int long_line_mode;
```
using 12 bytes to represent configuration options with 4 significant bits seems not extra optimized, but it's const any benefit would depend on implementation so I won't block, just leave a note.

ozersa replied:
Converted as uint8_t, thanks.

str4t0m commented on dts/bindings/w1/adi,max32-w1.yaml:
```diff
@@ -0,0 +1,64 @@
+    enum: [0, 1, 2]
+    description: |
+      Set this field to enable the external pullup.
+      0 - Pullup pin is active high when enabled.
+      1 - Pullup pin is active low when enabled.
```
Which configuration bits do you want to set with that property. And is available on all models?

str4t0m reviewed (changes requested):
Sorry for the long delay, I have a few comments.
Unfortunately, I have not yet ordered a board and I am not yet able to the test the implementation. May I ask if you have done any tests additionally to running the API test and the scanner sample?

ozersa commented:
> Sorry for the long delay, I have a few comments. Unfortunately, I have not yet ordered a board and I am not yet able to the test the implementation. May I ask if you have done any tests additionally to running the API test and the scanner sample?

Yes, I have tested with MAX31825 1W based temperature sensor too, please see [draft PR](https://github.com/zephyrproject-rtos/zephyr/pull/77550),

ozersa committed:
tests: drivers: w1: Move platform_allow in w1-serial

As per of str4t0m request, it is moved under
w1-serial section more detail in below link
https://github.com/zephyrproject-rtos/zephyr/pull/75009

Signed-off-by: Sadik Ozer <sadik.ozer@analog.com>

ozersa commented:
Rebased to fix conflictions.

ozersa commented:
@str4t0m please revisit when you get a chance.

ozersa commented:
Rebased to fix new conflicts that comes with https://github.com/zephyrproject-rtos/zephyr/pull/77002

ozersa commented:
@str4t0m I believe you are really busy, it is normal and fine, just a proposal to make progress for this PR, how about to merge it for now (This PR just add a driver in existing structure not includes a change that effect zephyr behavior) and incase of change/fix request we can revisit either zephyr test or drivers.

ozersa commented:
It is a kind reminder.

str4t0m commented on drivers/w1/w1_max32.c:
```diff
@@ -0,0 +1,207 @@
+	}
+
+	ret = pinctrl_apply_state(cfg->pctrl, PINCTRL_STATE_DEFAULT);
+	if (ret) {
```
NIT: `ret != 0`
